### PR TITLE
Add realtime site address validation

### LINKED ISSUE #9720: Better Error Messaging for Invalid URLs - Take 2
## Description

This is a follow up to https://github.com/wordpress-mobile/WordPress-Android/issues/8460

This feature has been [enhanced for WPiOS](https://github.com/wordpress-mobile/WordPress-iOS/issues/10294#issuecomment-486260221) beyond the initial solution proposed above.

The URL validation is now triggered by the user typing, and the "Next" button is not enabled until a valid URL is detected.

The URL validation error (pending editorial review) was changed to:

```
Enter a valid URL eg. example.com.
```

## iOS Demo:

Here's the iOS demo of the enhanced feature:

invalidURL

### PR DESCRIPTION
**Pull request by shiki**
Closes #9720.

This adds realtime validation for the site address in Login or Add Site. This is based on the iOS implementation described in https://github.com/wordpress-mobile/WordPress-iOS/issues/10294#issuecomment-486260221.

- The error is only shown if the user has not typed anything after 2 seconds
- The _Next_ button is only enabled if the URL is valid

To implement this, I moved all the site address logic to a new class, `LoginSiteAddressValidator`. The class encapsulates the validation, cleaning, and error reporting using debounce. A corresponding unit test, `LoginSiteAddressValidatorTest`, has also been added.

### Known Issues

The `LoginSiteAddressFragment` doesn't retain any error message shown below the text field after configuration changes. I didn't attempt to fix this. I thought it would be a huge change. 

### Adding a Site

1. Click on Sites → Switch Site.
2. Click on the + button at the top right.
3. Enter a site address.

### Logging in with a self-hosted site

1. On Login page, click on _Login_.
2. Click on _Log in by entering your site address_.
3. Enter a self-hosted site address.

On both scenarios, please test against the specs described above. Please also test that there are no regressions in logging in with a valid site. 

## Reviewing

Only 1 reviewer is needed but anyone can review.

### Editorial

We're hoping to get a review for the error message that is shown when the URL is invalid:

> Enter a valid URL eg. example.com.

## Release Notes

- [x] If there are user-facing changes, I have added an item to `RELEASE-NOTES.txt`.

## Tasks 

- [x] Review
- [ ] Merge if approved
- [ ] Merge changes back to `WordPress-Login-Flow-Android`
- [ ] Apply copy changes to iOS

### CODE REVIEW

File: libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginSiteAddressFragment.java [hunk 1]
```diff
@@ -359,7 +369,7 @@ public void onDiscoverySucceeded(OnDiscoveryResponse event) {
                 return;
             } else {
                 AppLog.e(T.API, "onDiscoveryResponse has error: " + event.error.name()
-                        + " - " + event.error.toString());
+                                + " - " + event.error.toString());
```
Comment (shiki) [inline libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/LoginSiteAddressFragment.java]: Auto-formatted. ¯\\\_(ツ)\_/¯

**Comment by benhuberman** [comment]
For the error message, I'd go with something that explains the error a bit more fully and avoids abbreviations (like e.g.). For example:
> Please enter a complete website address, like example.com.

This doesn't strike me as significantly longer than the current version, but let me know if it creates a length issue.

**Comment by shiki** [comment]
Thank you, @benhuberman. I've applied your suggestion. This is what it looks like now:

**Comment by benhuberman** [comment]
I like how it looks, @shiki.

I'm sorry I didn't catch this when reviewing this the first time, but I just saw that the first sentence on that screen could also use a small tweak -- the your/you sounds redundant (and it's strongly assumed that you'd only ever connect your own site). I'd go with this instead:
> Enter the address of the WordPress site you'd like to connect.

**Comment by shiki** [comment]
No worries at all @benhuberman. Updated:

**Pull request merged**